### DISCUSSION TITLE
'ORM Bulk UPDATE by Primary Key' not detecting primary key

### DISCUSSION THREAD

**Discussion by th02423** [sqlalchemy/sqlalchemy#10772] [Usage Questions] [answered]
Hello, 

I'm trying to update a table with a multi-column primary key using the syntax provided here: https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-update-by-primary-key

I have a table named `BulkUpdate` in a SQL Server database that looks like this:
![image](https://github.com/sqlalchemy/sqlalchemy/assets/153853747/daec1620-c091-4328-8a2b-eca6efbaae34)
And has this constraint: `CONSTRAINT PK_BulkUpdate PRIMARY KEY ([Name], [Number])`

To update this table I wrote this:
```
from sqlalchemy.sql.schema import Table
from sqlalchemy import create_engine, MetaData

engine = create_engine(..., fast_executemany=True) # Driver --> ODBC Driver 17 for SQL Server
meta = MetaData()
target_table = Table('BulkUpdate', meta, autoload_with=engine)

payload = [
    {'Name': 'Steve', 'Number': 1, 'Value': None},
    {'Name': 'Steve', 'Number': 2, 'Value': 16.27},
    {'Name': 'Bob', 'Number': 1, 'Value': 100.00}
]
with engine.begin() as conn:
    conn.execute(update(target_table), payload)
```

However, when I do this, a WHERE clause is never formed in the update query, even though the entire primary key is included in each entry that I include in the params for the `execute` statement, and this error is thrown:

> sqlalchemy.exc.IntegrityError: (pyodbc.IntegrityError) ('23000', "[23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Violation of PRIMARY KEY constraint 'PK_BulkUpdate'. Cannot insert duplicate key in object 'dbo.BulkUpdate'. The duplicate key value is (Steve, 1). (2627) (SQLExecDirectW); [23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The statement has been terminated. (3621)")
> [SQL: UPDATE [BulkUpdate] SET [Name]=?, [Number]=?, [Value]=?]
> [parameters: [('Steve', 1, None), ('Steve', 2, 16.27), ('Bob', 1, 100.0)]]

What is the proper way to invoke **bulk UPDATE by primary key mode** as stated in the docs?

**Comment by zzzeek** [comment, marked as answer]
ORM update by primary key is only enabled when you use an ORM model, not a Table object, as the subject of the update.

to get an autoloaded version of an ORM model, use one of the approaches described at https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapping-declaratively-with-reflected-tables

**Reply by lpratas-nexa** [reply]
Is there any particular reason for this? Bulk insert seems to work with Table objects, and I feel like this distinction is not made clear enough in the documentation.

**Reply by CaselIT** [reply]
when working with table and more generally with core objects, the idea is that what you write using the statement api is what's get executed. `update(target_table)` has no where clause so none is rendered.
If you specify it, it works as intended:
```py
payload = [
    {"Name_": "Steve", "Number_": 1, "Value": None},
    {"Name_": "Steve", "Number_": 2, "Value": 16.27},
    {"Name_": "Bob", "Number_": 1, "Value": 100.00},
]
with engine.begin() as conn:
    conn.execute(
        update(target_table).where(
            target_table.c.Name == bindparam("Name_"),
            target_table.c.Number == bindparam("Number_"),
        ),
        payload,
    )
```

bulk insert does not reply on a where clause being present, so sqlalchemy is not adding part of the sql statement
